src/strongvelope/strongvelope.cpp

```diff
@@ -738,18 +805,50 @@ void ProtocolHandler::rsaDecrypt(const StaticBuffer& data, Buffer& output)
 }
 
 promise::Promise<std::pair<MsgCommand*, KeyCommand*>>
-ProtocolHandler::msgEncrypt(Message* msg, MsgCommand* msgCmd)
+ProtocolHandler::msgEncrypt(Message* msg, const SetOfIds &recipients, MsgCommand* msgCmd)
 {
     assert(msg->keyid == msgCmd->keyId());
-    if ((msg->keyid == CHATD_KEYID_INVALID)
-     || (msg->keyid == CHATD_KEYID_UNCONFIRMED)) //we have to use the current send key
+    if ((msg->keyid == CHATD_KEYID_INVALID)         // keyid has not been assigned yet (NEWMSG)
+     || (msg->keyid == CHATD_KEYID_UNCONFIRMED))    // or it was, but not yet confimed (MSGUPDX, replayed NEWMSG)
     {
-        if (!mCurrentKey || mParticipantsChanged) // create a new key and prepare the KeyCommand
+        assert(msgCmd->opcode() != OP_MSGUPD);  // edits always have a confirmed keyid
+
+        if (mCurrentKey && mCurrentKeyParticipants == *mParticipants)
+        {
+            // use current key, whether it's unconfirmed (in-flight) or already confirmed
+            assert(mCurrentKey);
+            assert(mCurrentKeyId != CHATD_KEYID_INVALID);
+            msg->keyid = mCurrentKeyId;
+            msgCmd->setKeyId(mCurrentKeyId);
+            msgEncryptWithKey(*msg, *msgCmd, *mCurrentKey);
+            return std::make_pair(msgCmd, (KeyCommand*)nullptr);
+        }
+        else    // msg for different set of recipients or no current key available
         {
+            // edits of unconfirmed new messages should reuse the same key than original message
+            if (msgCmd->opcode() == OP_MSGUPDX)
+            {
+                std::shared_ptr<SendKey> unconfirmedKey;
+                for (auto& it: mUnconfirmedKeys)
+                {
+                    if (it.first == recipients)
```

Comment:
This might cause issues in the case of:
```K1 M1 M2 K3 M3 K4 M4 M4updx```
where the set of participants for `K1` is the same than for `K4`. Hence, the `M4updx` will use the `K1` instead of the right one, `K4`.
Solution: attach to the `mUnconfirmedKeys` map the list of `msgxid` using the unconfirmed key when a message uses the key. It should be done as well in the `loadUnconfirmedKeysFromDb()`.